### ISSUE TITLE
Windows executor can't find slack channel

### ISSUE THREAD

**Issue by ashmortar** [CircleCI-Public/slack-orb#217]
## Orb version:
4.1.3

## What happened:

I run a notification step running on fail at the end of each job. The step works fine except in the windows executor. Where I get the message: 

```bash
Sending to Slack Channel: ***********
{
  "ok": false,
  "error": "channel_not_found"
}
```

On Linux machine and Macos images the notification is sent correctly.

## Expected behavior:

All platforms work the same.

## Additional Information:

The slack channel I am posting to does contain a `-` in the name. I haven't tried testing on a different channel yet but that is my next step.

**Comment by gmemstr** [comment]
@ashmortar To confirm, you are specifying bash as a shell when running under Windows?

**Comment by ashmortar** [comment]
Yes

**Comment by ashmortar** [comment]
here is the full error message being output:

```
Posting Status
cat: /etc/issue: No such file or directory
cat: /etc/issue: No such file or directory
cat: /etc/issue: No such file or directory
sed: -e expression #1, char 8: unterminated `s' command
Sending to Slack Channel: ***********
{
  "ok": false,
  "error": "channel_not_found"
}
```

**Comment by Kurt-von-Laven** [comment]
You can use build artifacts to pass the outcome of a job from a Windows workflow to a Linux workflow and then trick the Slack orb into reporting the status of the Windows workflow from the Linux workflow by overriding the environment variables it reads from. This is a very ugly solution, but I personally prefer it to having no Slack notifications at all on Windows. Another option would be to ditch the Slack orb altogether and hit the Slack API directly. Any feedback folks have on improving this workaround is quite welcome.

```
version: 2.1
orbs:
  slack: circleci/slack@4.4.0
  windows: circleci/windows@2.4.0
jobs:
  build:
    executor: windows/default
    steps:
      - run:
          name: Make directory for job metadata.
          command: mkdir C:/tmp/job_metadata
          shell: bash.exe
      - run:
          name: Write current job URL.
          command: echo "$CIRCLE_BUILD_URL" > C:/tmp/job_metadata/build_url.txt
          shell: bash.exe
...
      - run:
          when: on_fail
          name: Write job exit code 1 for failure.
          command: echo 1 > C:/tmp/job_metadata/exit_code.txt
          shell: bash.exe
      - run:
          when: on_success
          name: Write job exit code 0 for success.
          command: echo 0 > C:/tmp/job_metadata/exit_code.txt
          shell: bash.exe
      - store_artifacts:
          path: /tmp/job_metadata
          destination: build-job-metadata
  notify:
    docker:
      - image: cimg/base:stable
    resource_class: small
    steps:
      - run:
          name: Download any job metadata artifacts for latest completed job on this branch.
          command: >
            curl
            --header "Content-Type: application/json"
            --header "Accept: application/json"
            --header "Circle-Token: $CIRCLE_TOKEN"
            "https://circleci.com/api/v1.1/project/<< pipeline.project.type >>/$CIRCLE_PROJECT_USERNAME/$CIRCLE_PROJECT_REPONAME/latest/artifacts?branch=$CIRCLE_BRANCH&filter=completed" |
            grep --only-matching 'https://[^"]*build-job-metadata[^"]*' |
            xargs -I url wget --header "Circle-Token: $CIRCLE_TOKEN" url
      - run:
          name: Overwrite CIRCLE_JOB environment variable with build.
          command: echo export CIRCLE_JOB=build >> $BASH_ENV
      - run:
          name: Overwrite CIRCLE_BUILD_URL environment variable from prior job metadata.
          command: echo export CIRCLE_BUILD_URL=$(<build_url.txt) >> $BASH_ENV
      - run:
          name: Exit with code from prior job metadata.
          command: exit $(<exit_code.txt)
      - slack/notify:
          event: fail
          template: basic_fail_1
      - slack/notify:
          event: pass
          template: basic_success_1
workflows:
  build:
    jobs:
      - build
  # Take advantage of the fact that free CircleCI accounts are limited to
  # running one workflow at a time. If you are on a paid account, these
  # workflows will instead run in parallel. Consider submitting a pull request
  # to the Slack orb to use the bash.exe shell on Windows, using
  # https://github.com/eddiewebb/circleci-queue, or using the CircleCI API to
  # wait for the build workflow to complete.
  notify:
    jobs:
      - notify:
          context:
            # a context that defines the CIRCLE_TOKEN environment variable
            # to be a CircleCI API token for your project with Admin scope
            - circleci-api-admin
            # a context that defines the SLACK_ACCESS_TOKEN and SLACK_DEFAULT_CHANNEL
            # environment variables as specified by
            # https://github.com/CircleCI-Public/slack-orb#readme
            - slack
```

**Comment by shortstack** [comment]
I am also having a similar issue

```
orbs:
  slack: circleci/slack@4.4.4
```

Mac and Ubuntu work fine, Windows does this with the same configuration
```
Posting Status
cat: /etc/issue: No such file or directory
cat: /etc/issue: No such file or directory
cat: /etc/issue: No such file or directory
sed: -e expression #1, char 8: unterminated `s' command
Sending to Slack Channel: #REDACTED
Slack API returned an error message:
"invalid_arguments"


View the Setup Guide: https://github.com/CircleCI-Public/slack-orb/wiki/Setup
CircleCI received exit code 0
```


windows config machine:
```
  build-windows:
    machine:
      image: windows-server-2019-vs2019:stable
      resource_class: windows.medium
      shell: bash.exe
```

**Comment by Kurt-von-Laven** [comment]
@shortstack, the Slack orb doesn't support Windows. Take a look at my workaround.

**Comment by shortstack** [comment]
trying it now :( just feels so hacky. thanks @Kurt-von-Laven !

**Comment by Kurt-von-Laven** [comment]
@shortstack, certainly can't disagree that it feels hacky. I recommend using [circleci-queue](https://github.com/eddiewebb/circleci-queue) or the CircleCI API to wait for the build to complete. I'm sure others will appreciate it if you post whatever ends up working for you.

**Comment by mkarp** [comment]
I've tried debugging this issue and also couldn't find a good solution yet. 

It seems like on Windows VMs at some point there's an issue with escaping backslashes in this command `sed 's/\\/\\\\/g'`, but it's unclear to me why and when.

Here's some findings from a debugging session.

* I've enabled tracing like so: `echo "set -o xtrace" >> $BASH_ENV`

* Both on a Linux VM and on a Windows VM, the inlined `notify.sh` bash script seems to be loaded just fine, notice that `sed 's/\\/\\\\/g'` is correctly printed in both cases:

Linux (Docker):
```bash
#!/bin/bash -eo pipefail
BuildMessageBody() {
...
        CUSTOM_BODY_MODIFIED=$(echo "$CUSTOM_BODY_MODIFIED" | sed 's/\\/\\\\/g'
```

Windows:
```bash
#!bash.exe
BuildMessageBody() {
...
        CUSTOM_BODY_MODIFIED=$(echo "$CUSTOM_BODY_MODIFIED" | sed 's/\\/\\\\/g'
```

* However, then during the run on Windows the backslashes are already escaped:

Linux (Docker):
```bash
...
++ echo '{
  "text": "CircleCI job failed.",
  "blocks": [
    {
      "type": "header",
...
}'
++ sed 's/\\/\\\\/g'
...
```

Windows:
```bash
...
++ echo '{
  "text": "CircleCI job failed.",
  "blocks": [
    {
      "type": "header",
...
}'
++ sed 's/\/\\/g'
sed: -e expression #1, char 8: unterminated `s' command
...
```

I've run another experiment on Circle's Windows VM. I've put the contents of a Slack message template to `test.json`, then created a script `test.sh` with the following contents: `cat test.json | sed 's/\\/\\\\/g'`, and it runs as expected:

```bash
PS C:\Users\circleci\project> bash test.sh
bash test.sh
+ cat test.json
+ sed 's/\\/\\\\/g'
{
  "text": "CircleCI job failed.",
  "blocks": [
    {
      "type": "header",
...
}
```

But running the `sed` command inline doesn't work:

```bash
PS C:\Users\circleci\project> bash -c "cat test.json | sed 's/\\/\\\\/g'"
bash -c "cat test.json | sed 's/\\/\\\\/g'"
+ cat test.json
+ sed 's/\/\\/g'
sed: -e expression #1, char 8: unterminated `s' command
```

So perhaps because the bash script `notify.sh` is inlined in the Orb config, backslashes are escaped prematurely?

**Comment by Kurt-von-Laven** [comment]
The Slack orb doesn't support Windows. It tries to execute Bash scripts as if they were PowerShell scripts if I recall correctly. Please see my workaround above.

**Comment by KyleTryon** [comment]
Hey folks, sorry we couldnt catch this earlier, @Kurt-von-Laven is correct. The issue is with the version of `sed` installed in Windows. Because of the unique way we access "linux" on windows, we have not yet found a viable solution to installing sed, as it is part of another installation in the Windows build process. We have the team who builds our Windows VM looking into the issue.

**Comment by mkarp** [comment]
@KyleTryon Thank you for your feedback!

> The issue is with the version of sed installed in Windows.

While it might be true, it doesn't appear like it for the error reported in this thread. I've tried [several experiments](https://github.com/CircleCI-Public/slack-orb/issues/217#issuecomment-987060538) and the sed command from the Slack orb worked similar to how it works on Docker/Mac VMs. The issue seems to be that the backslashes are escaped prior to being executed in bash on Windows:

This fails with `sed: -e expression #1, char 8: unterminated 's' command`: `cat message_template.json | sed 's/\\/\\\\/g'"`, the command is copied from [Slack orb's notify.sh](https://github.com/CircleCI-Public/slack-orb/blob/master/src/scripts/notify.sh#L9)

This works: `cat message_template.json | sed 's/\\\\/\\\\\\\\/g'`

I'm treating the CircleCI Windows VM as a black box at the moment (is it open source?) but it feels almost like it's executing bash commands with `bash -c command`, and that makes the backslashes being unnecessarily escaped. However, this should work `echo command > command.sh && bash command.sh`.

**Comment by EricRibeiro** [comment]
As mentioned in the [wiki's FAQ](https://github.com/CircleCI-Public/slack-orb/wiki/FAQ) section, the slack-orb does not work with Windows. To quote it:

> Alternative integrations will be required for Windows based builds. Consider connecting our [CircleCI Webhooks](https://circleci.com/docs/2.0/webhooks/) with an automation service which connects to Slack.

**Comment by EricRibeiro** [comment]
Hey, @ashmortar 👋 

I know it's been a while, and sorry for that, but support for Windows was implemented in [v4.12.0](https://github.com/CircleCI-Public/slack-orb/releases/tag/v4.12.0). If you want to give it a shot, make sure to choose `bash.exe` as your [shell in the Windows executor](https://circleci.com/docs/using-windows/#specifying-a-shell-with-the-windows-executor).

Let me know how it goes!
